tempesta_fw/apm.c

```diff
@@ -0,0 +1,734 @@
+}
+
+/*
+ * Consolidate and merge all entries in @rnglst for correct percentile
+ * calculation.
+ *
+ * The ranges may be changed during the stats updates. Ranges within
+ * a single entry are guaranteed to not intersect. However, ranges in
+ * different entries may intersect. The intersecting ranges need to be
+ * merged carefully and accurately.
+ *
+ * The algorithm is that all ranges are split into separate buckets.
+ * Each bucket is a response time value and the number of hits. All
+ * buckets are sorted by the response time value, so that buckets with
+ * identical or close time values are placed together. This makes for
+ * correct calculation of percentiles by consecutive summing of the
+ * number of hits in these buckets.
+ */
+static int
+tfw_apm_calc_cmp(const void *a, const void *b)
+{
+	const TfwPcntCalc *x = a, *y = b;
+
+	if (x->rtmval > y->rtmval)
+		return 1;
+	if (x->rtmval < y->rtmval)
+		return -1;
+	return 0;
+}
+
+static int
+tfw_apm_rngmerge(TfwPcntCalc *pcalc, TfwPcntRanges **rnglst, size_t rlsz)
+{
+	int i, r, b, pcsz;
+	TfwPcntCtl pctl[TFW_STATS_RANGES];
+
+	for (i = 0, pcsz = 0; i < rlsz; ++i) {
+		spin_lock(&tfw_stats_sa_guard);
+		for (r = 0; r < TFW_STATS_RANGES; ++r)
+			pctl[r].atomic = rnglst[i]->ctl[r].atomic;
+		spin_unlock(&tfw_stats_sa_guard);
+
+		for (r = 0; r < TFW_STATS_RANGES; ++r) {
+			for (b = 0; b < TFW_STATS_BCKTS; ++b) {
+				if (!atomic_read(&rnglst[i]->cnt[r][b]))
+					continue;
+				pcalc[pcsz].intrvl = i;
+				pcalc[pcsz].range = r;
+				pcalc[pcsz].bucket = b;
+				pcalc[pcsz++].rtmval =
+					pctl[r].begin + (b << pctl[r].order);
+			}
+		}
+	}
+	sort(pcalc, pcsz, sizeof(TfwPcntCalc), tfw_apm_calc_cmp, NULL);
+
+	return pcsz;
+}
+
+static int
+tfw_apm_rnglst(TfwApmRBuf *rbuf, TfwPcntRanges **rnglst)
+{
+	TfwApmRBufEntry *rbent;
+	unsigned long jmax = 0, jtmwstart;
+	int i, imax = 0, rlsz = 0, i_rlsz;
+
+	/*
+	 * Find the entry with the maximim @jtstamp. That's the most
+	 * recent entry. The entries that precede the current entry
+	 * are used in the calculation of percentiles;
+	 */
+	for (i = 0; i < rbuf->rbentsz; ++i) {
+		if (jmax < rbuf->rbent[i].jtstamp) {
+			jmax = rbuf->rbent[i].jtstamp;
+			imax = i;
+		}
+	}
+	/*
+	 * If the entry with the maximum @jstamp is the current active
+	 * entry, then add it to the list of entries. Take care of the
+	 * circular nature of the buffer.
+	 */
+	rlsz = tfw_apm_tmwscale;
+	if (rbuf->rbent[imax].jtstamp >= jiffies - tfw_apm_jtmintrvl)
+		rlsz++;
+	imax += rbuf->rbentsz;
+	/*
+	 * Collect entries that are within the defined time window.
+	 * Most recent entries are put first in the list, older entries
+	 * are put last in the list. Take care of periods of inactivity.
+	 */
+	jtmwstart = jiffies - tfw_apm_jtmwindow;
+	for (i = 0, i_rlsz = 0; i < rlsz; ++i, --imax) {
+		rbent = &rbuf->rbent[imax % rbuf->rbentsz];
+		if (rbent->jtstamp && (rbent->jtstamp >= jtmwstart))
+			rnglst[i_rlsz++] = &rbent->pcntrng;
+	}
+
+	return i_rlsz;
+}
```

Comment:
The function is required to track periods of inactivity only and I believe we can just remove it. Period of inactivity is technically just zeroed range. So lets just zero previous range on timer. The only timer should do statistic calculation, the reasonable value is `HZ / 20`. Since the timer runs more frequently than ranges are changed it shall track the last statistic update timestamp and if it determines that previous range wasn't updated due to period if inactivity, the range must be zeroed (vectorized `memset()` will do this quickly). Only after the check and zeroing the timer function should do statistics recalculation.